ngraph/src/ngraph/node.cpp

```diff
@@ -800,7 +802,14 @@ bool Node::match_value(pattern::Matcher* matcher,
 bool Node::match_node(pattern::Matcher* matcher, const Output<Node>& graph_value)
 {
     matcher->add_node(graph_value);
-    if (graph_value.get_node_shared_ptr()->get_type_info() == get_type_info() &&
+    // Check if a type of a given node, which produces graph_value, matches the type of `this` node
+    // or `this` node type is an ancestor of that node type. It is not the exact matching, types of
+    // the nodes
+    // may not match, but they are connected by the inheritance relation.
+    // Not exact matching allows using base classes in the patterns and successfully matching such
+    // patterns
+    // with sub-graph of descent nodes types.
+    if (graph_value.get_node_shared_ptr()->get_type_info().is_castable(get_type_info()) &&
```

Comment:
@GlebKazantaev, please review this particular piece in order to understand if it is possible to "fast-track" it outside to avoid checking for each matcher node.

Reply:
Modified "fast-track" code in GraphRewrite::run_on_function, covered by tests.